### PR TITLE
KAFKA-5233 follow up

### PR DESCRIPTION
**Pull request by mihbor**

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6831/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6817/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by guozhangwang** [comment]
@dguy could you take a look?

**Review by guozhangwang — COMMENTED** [review]
Thinking a bit about this. Maybe we could support `schedule` in `MockProcessorContext` which maintains a map from processor node to a list of its registered punctuate functions. Then in `StreamTestDriver.punctuate()` traversing through the processors in topology and call their punctuate functions?

BTW I realized that with KIP-138 we are not allowing punctuate functions to return a value any more. Should this be a problem?

**Comment by peoplemerge** [comment]
@mihbor punctuate() is working great in our unit tests.  However, at scale, when I call context.forward, we're getting:

```
java.lang.NullPointerException: null
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:83) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at com.ticketmaster.service.offersstreamconnector.BurstReducingProcessor.punctuate(BurstReducingProcessor.java:76) ~[classes!/:na]
	at org.apache.kafka.streams.processor.internals.ProcessorNode$4.run(ProcessorNode.java:133) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at org.apache.kafka.streams.processor.internals.StreamsMetricsImpl.measureLatencyNs(StreamsMetricsImpl.java:187) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at org.apache.kafka.streams.processor.internals.ProcessorNode.punctuate(ProcessorNode.java:136) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at org.apache.kafka.streams.processor.internals.StreamTask.punctuate(StreamTask.java:238) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at org.apache.kafka.streams.processor.internals.PunctuationQueue.mayPunctuate(PunctuationQueue.java:50) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at org.apache.kafka.streams.processor.internals.StreamTask.maybePunctuateSystemTime(StreamTask.java:540) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at org.apache.kafka.streams.processor.internals.StreamThread$2.apply(StreamThread.java:749) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at org.apache.kafka.streams.processor.internals.StreamThread.performOnStreamTasks(StreamThread.java:1474) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at org.apache.kafka.streams.processor.internals.StreamThread.maybePunctuateSystemTime(StreamThread.java:739) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:567) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
	at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:527) ~[kafka-streams-0.11.0.0+efb060c57f05d1d586bb14c016b0187c60f8e994-SNAPSHOT.jar!/:na]
```
Note - this is #3055 cherrypicked against 0.11.0.0.  If you think trunk up to efb060 is better I can try that.

Further - I can't prove 100% it's related to #3055; suggestions appreciated!

File: streams/src/test/java/org/apache/kafka/test/KStreamTestDriver.java [hunk 1]
```diff
@@ -189,7 +190,23 @@ private ProcessorNode sourceNodeByTopicName(final String topicName) {
         return topicNode;
     }
 
-    public void punctuate(final long timestamp) {
+    public void punctuate(final long timestamp, Punctuator punctuator) {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/test/KStreamTestDriver.java:193]: This is going to call punctuate on every node - right? Which is not actually correct, I think something like @guozhangwang suggested might be a better approach.

**Comment by guozhangwang** [comment]
Adding @mjsax as a reviewer as well since we are working on replacing the internal test driver with the new public test driver now.

**Comment by mjsax** [comment]
@mihbor Are you aware of KIP-247 ? I am not sure if we should update `KStreamTestDriver` as we plan to remove it anyway...

**Comment by mihbor** [comment]
Thanks @mjsax, I wouldn't be able to look into this until March, so it's good news that it's going to be removed anyway :)

**Comment by mjsax** [comment]
Sound like we can close this PR? If there is anything missing in KIP-247, we should add it to the new public `TopologyTestDriver` IMHO. WDYT?

**Comment by guozhangwang** [comment]
@mjsax Sounds good to me. In fact I was cc'ing you just as an FYI.

**Pull request closed without merging**